clippy_lints/src/unconventional_macro_braces.rs

```diff
@@ -0,0 +1,146 @@
+use clippy_utils::{diagnostics::span_lint_and_help, in_macro, is_direct_expn_of, source::snippet_opt};
+use if_chain::if_chain;
+use rustc_ast::ast;
+use rustc_lint::{EarlyContext, EarlyLintPass};
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use rustc_span::Span;
+
+declare_clippy_lint! {
+    /// **What it does:** Checks that common macros are used with consistent bracing.
+    ///
+    /// **Why is this bad?** This is mostly a consistency lint although using () or []
+    /// doesn't give you a semicolon in item position, which can be unexpected.
+    ///
+    /// **Known problems:**
+    /// None
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// vec!{1, 2, 3};
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// vec![1, 2, 3];
+    /// ```
+    pub UNCONVENTIONAL_MACRO_BRACES,
+    nursery,
+    "check consistent use of braces in macro"
+}
+
+#[derive(Clone, Copy, Debug)]
+struct MacroMatcher {
+    name: &'static str,
+    braces: (&'static str, &'static str),
+}
+
+const BRACES: &[(&str, &str)] = &[("(", ")"), ("{", "}"), ("[", "]")];
+
+const MACROS_TO_BRACES: &[MacroMatcher] = &[
+    MacroMatcher {
+        name: "print",
+        braces: ("(", ")"),
+    },
+    MacroMatcher {
+        name: "println",
+        braces: ("(", ")"),
+    },
+    MacroMatcher {
+        name: "eprint",
+        braces: ("(", ")"),
+    },
+    MacroMatcher {
+        name: "eprintln",
+        braces: ("(", ")"),
+    },
+    MacroMatcher {
+        name: "write",
+        braces: ("(", ")"),
+    },
+    MacroMatcher {
+        name: "format",
+        braces: ("(", ")"),
+    },
+    MacroMatcher {
+        name: "format_args",
+        braces: ("(", ")"),
+    },
+    MacroMatcher {
+        name: "vec",
+        braces: ("[", "]"),
+    },
+    MacroMatcher {
+        name: "quote",
+        braces: ("{", "}"),
+    },
```

Comment:
I think all of the above should be clear-cut but `quote!` is also used as `quote!()` a lot. Maybe this lint could be warn-by-default with `quote` removed (which users could then re-add as a config item)?